### PR TITLE
Explicitly raise InvalidResponse if commands and responses counts don't match

### PR DESCRIPTION
**Pull request by byroot**
As discussed in https://github.com/redis/redis-rb/pull/509

The idea here is to fail loudly if such unlikely race condition happened.

@djanowski is this more acceptable to you?

### CODE REVIEW

File: lib/redis/errors.rb [hunk 1]
```diff
@@ -14,6 +14,10 @@ def initialize(reply_type)
     end
   end
 
+  # Raised when the response from the server is unparsable
+  class InvalidResponse < BaseError
+  end
+
```
Comment (djanowski) [inline lib/redis/errors.rb]: `ProtocolError`?

File: lib/redis/pipeline.rb [hunk 2]
```diff
@@ -58,6 +58,10 @@ def without_reconnect(&blk)
     end
 
     def finish(replies, &blk)
+      unless replies.size == futures.size
+        raise InvalidResponse, "#{futures.size} commands were enqueued, but only #{replies.size} responses were recieved"
+      end
```
Comment (djanowski) [inline lib/redis/pipeline.rb]: Typo.

File: lib/redis/errors.rb (hunk 1 above)
Comment (byroot) [inline lib/redis/errors.rb]: I wanted to use ProtocolError initially, but it's constructor is customized. I can refactor it though.

File: lib/redis/pipeline.rb [hunk 3]
```diff
@@ -71,7 +75,7 @@ def finish(replies, &blk)
 
     class Multi < self
       def finish(replies)
-        exec = replies.last
+        exec = replies.pop
```
Comment (djanowski) [inline lib/redis/pipeline.rb]: If possible, let's not mutate an argument.

Comment (byroot) [inline lib/redis/pipeline.rb]: I can copy the array instead if you think it's best.

**Comment by byroot** [comment]
I fixed the typo, and refactored `ProtocolError` so I can reuse it.

I honestly think the `pop` is simpler and better than the copy. I mean I get your point about not mutating arguments, but in this case it's a method responsible for parsing a response. So I really don't expect any further usage of that object.

That being said I don't think super strongly about it. So I can totally do a `slice` instead. Just let me know.

**Comment by byroot** [comment]
We no longer need this because pipelines are serialized eagerly.

**Pull request closed without merging**